File: Frontend/src/utils/trl.js
export const TRL_LABELS = {
  1: 'Basic principles observed',
  2: 'Technology concept formulated',
  3: 'Experimental proof of concept',
  4: 'Technology validated in lab',
  5: 'Technology validated in relevant environment',
  6: 'Technology demonstrated in relevant environment',
  7: 'System prototype demonstrated in operational environment',
  8: 'System complete and qualified',
  9: 'Actual system proven in operational environment',
}

/** RAG badge colours: red = early, amber = middle, green = TRL 7+. */
export const TRL_COLORS = {
  1: 'bg-red-50 text-red-700',
  2: 'bg-red-50 text-red-700',
  3: 'bg-red-50 text-red-700',
  4: 'bg-amber-50 text-amber-800',
  5: 'bg-amber-50 text-amber-800',
  6: 'bg-amber-50 text-amber-800',
  7: 'bg-emerald-50 text-emerald-800',
  8: 'bg-emerald-50 text-emerald-800',
  9: 'bg-emerald-50 text-emerald-800',
}

/** Investor-facing maturity bands — plain language first, TRL as secondary detail. */
export const TRL_FILTERS = [
  {
    key: 'all',
    label: 'All stages',
    description: 'Every investable project',
    min: null,
    max: null,
  },
  {
    key: '4-5',
    label: 'Early validation',
    description: 'Validated in the lab or a relevant clinical setting',
    trlRange: 'TRL 4–5',
    min: 4,
    max: 5,
  },
  {
    key: '6-7',
    label: 'Clinical demonstration',
    description: 'Demonstrated in real-world healthcare environments',
    trlRange: 'TRL 6–7',
    min: 6,
    max: 7,
  },
  {
    key: '8-9',
    label: 'Market ready',
    description: 'Complete, qualified, and proven in live use',
    trlRange: 'TRL 8–9',
    min: 8,
    max: 9,
  },
]

export function trlStageBand(trl) {
  return TRL_FILTERS.find((item) => item.min && trl >= item.min && trl <= item.max) || null
}

export function trlStageBandLabel(trl) {
  return trlStageBand(trl)?.label || trlFullLabel(trl)
}

export function trlShortLabel(trl) {
  return `TRL ${trl}`
}

export function trlFullLabel(trl) {
  return TRL_LABELS[trl] || `Technology readiness level ${trl}`
}

export function trlBadgeClass(trl) {
  return TRL_COLORS[trl] || TRL_COLORS[4]
}

export function matchesTrlFilter(trl, filterKey) {
  const filter = TRL_FILTERS.find((item) => item.key === filterKey)
  if (!filter || filter.key === 'all') return true
  return trl >= filter.min && trl <= filter.max
}

File: Backend/app/trl.py
"""Technology Readiness Level (TRL 1–9) helpers for RHIP innovation projects."""

TRL_LABELS: dict[int, str] = {
    1: "Basic principles observed",
    2: "Technology concept formulated",
    3: "Experimental proof of concept",
    4: "Technology validated in lab",
    5: "Technology validated in relevant environment",
    6: "Technology demonstrated in relevant environment",
    7: "System prototype demonstrated in operational environment",
    8: "System complete and qualified",
    9: "Actual system proven in operational environment",
}


def trl_label(trl: int) -> str:
    return TRL_LABELS.get(trl, f"TRL {trl}")


def trl_short_label(trl: int) -> str:
    return f"TRL {trl}"


def readiness_from_trl(trl: int) -> str:
    """Map TRL to legacy readiness band for government/pipeline filters."""
    if trl <= 3:
        return "early"
    if trl <= 5:
        return "feasibility"
    if trl <= 7:
        return "clinical"
    return "commercial"


def trl_from_readiness(readiness: str, stage: int = 5) -> int:
    """Derive TRL from legacy readiness + stage when migrating seed data."""
    base = {
        "early": 3,
        "feasibility": 5,
        "clinical": 6,
        "commercial": 8,
    }.get(readiness, 5)
    if stage >= 8 and base < 8:
        return 8
    if stage >= 7 and base < 7:
        return 7
    if stage >= 6 and base < 6:
        return 6
    return base
